### ISSUE TITLE
Recovering from keyword identifier mentions raw identifier that never appeared in the source code later

### ISSUE THREAD

**Issue by ChayimFriedman2** [rust-lang/rust#97854]
```rust
fn main() {
    let while;
}
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=92d99a1b9f37c71ce9f5a5b26f511489
```
error: expected identifier, found keyword `while`
 --> src/main.rs:2:9
  |
2 |     let while;
  |         ^^^^^ expected identifier, found keyword
  |
help: escape `while` to use it as an identifier
  |
2 |     let r#while;
  |         ++

error[E0282]: type annotations needed
 --> src/main.rs:2:9
  |
2 |     let while;
  |         ^^^^^ consider giving `r#while` a type
```
The second error refers to `r#while` while we wrote `while`, because we recover as if we've written a raw identifier. Feel free to close if this is not a problem.

**Comment by compiler-errors** [comment]
This is not ideal, but pretty difficult to fix IMO.

**Comment by ChayimFriedman2** [comment]
Yes, I also think that, but maybe we can see if the identifier was written raw. I'll take a look later.

**Comment by compiler-errors** [comment]
I'm hesitant to suggest fixing this, because it would require fixing (and thus complicating)  _every_ case where we rely on being able to print the the ident of an variable expression in a suggestion.

The only possible fix I would recommend investigating would be suggesting `r#while`, but keeping the variable's own ident as `while` instead of replacing it with the recovery `r#while`. However, the issue with this is that other suggestions that rely on the variable name being valid in suggestions (i.e. "did you mean the variable named `while`" suggestion for a typo) would then break.

I personally prefer the state of the world as it is now, I think.

**Comment by ChayimFriedman2** [comment]
> The only possible fix I would recommend investigating would be suggesting `r#while` in this message, but keeping the variable's own ident as `while` instead of replacing it with the recovered `r#while`.

This is what currently happens: https://github.com/rust-lang/rust/blob/47aee31b2a89cb7de97d779869a30b046632b6af/compiler/rustc_parse/src/parser/mod.rs#L535. I guess that the printing machinery automatically adds `#`.

**Comment by compiler-errors** [comment]
Ah, makes sense. Seems like someone else also thought that we should render `r#while` for the ident always :stuck_out_tongue_winking_eye:

**Comment by ChayimFriedman2** [comment]
Seems to me more like someone that though that if you have an identifier as a keyword, it was obviously escaped. Can't blame them :)

**Comment by veera-sivarajan** [comment]
Output on `rustc 1.79.0`:
```
error: expected identifier, found keyword `while`
 --> <source>:3:9
  |
3 |     let while;
  |         ^^^^^ expected identifier, found keyword
  |
help: escape `while` to use it as an identifier
  |
3 |     let r#while;
  |         ++

error: aborting due to 1 previous error

```
